File: LeetCode/c++/0658.cpp
#include <vector>
#include <cmath>
#include <iostream>

using namespace std;


class Solution {
    public:
        vector<int> findClosestElements(vector<int>& arr, int k, int x) {
            /*
            Time:  O(n)
            Space: O(k)
            */
            int start = 0, end = arr.size() - 1;
            while (end - start >= k) {
                if (abs(arr[start] - x) <= abs(arr[end] - x)) end--;
                else start++;
            }
            vector<int> result;
            for (int i = start; i <= end; i++) result.push_back(arr[i]);
            return result;
        }
    };


int main() {
    vector<int> arr = {1, 2, 3, 4, 5};
    Solution solution;
    vector<int> result = solution.findClosestElements(arr, 4, 3);
    
    for (int num : result) {
        cout << num << " ";
    }

    return 0;
}

File: LeetCode/golang/0658.go
func Abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}

func findClosestElements(arr []int, k int, x int) []int {
	/**
	* Time:  O(n-k)
	* Space: O(1)
	*/
	start, end := 0, len(arr) - 1

	for end - start + 1 > k {
		if Abs(arr[start] - x) <= Abs(arr[end] - x) {
			end--
		} else {
			start++
		}
	}
	return arr[start:end+1]
}